src/robot.cpp

```diff
@@ -37,32 +88,78 @@ Vector7d Robot::currentJointVelocities() {
   return currentJointState().velocity();
 }
 
+franka::Model* Robot::lazyModel() {
+  if (!model_) {
+    model_ = std::make_unique<franka::Model>(loadModel());
+  }
+  return model_.get();
+}
+
+std::array<double, 16>& Robot::lazy_F_T_EE() {
+  if (!F_T_EE_) {
+    F_T_EE_ = std::make_unique<std::array<double, 16>>(state().F_T_EE);
+  }
+  return *F_T_EE_;
+}
+
+std::array<double, 16>& Robot::lazy_EE_T_K() {
+  if (!EE_T_K_) {
+    EE_T_K_ = std::make_unique<std::array<double, 16>>(state().EE_T_K);
+  }
+  return *EE_T_K_;
+}
+
 Affine Robot::forwardKinematics(const Vector7d &q) {
-  return Kinematics::forward(q);
+  std::array<double, 7> q_array;
+  for (size_t i = 0; i < 7; ++i) { q_array[i] = q[i]; }
+
+  std::array<double, 16> pose = lazyModel()->pose(franka::Frame::kEndEffector, q_array, lazy_F_T_EE(), lazy_EE_T_K());
+  return Affine(Eigen::Matrix4d::Map(pose.data()));
 }
 
-Vector7d Robot::inverseKinematics(const Affine &target, const Vector7d &q0) {
-  Vector7d result;
+Jacobian Robot::jacobian(const Vector7d &q) {
+  std::array<double, 7> q_array;
+  for (size_t i = 0; i < 7; ++i) { q_array[i] = q[i]; }
 
-  Eigen::Vector3d angles = Euler(target.rotation()).angles();
-  Eigen::Vector3d angles_norm;
-  angles_norm << angles[0] - M_PI, M_PI - angles[1], angles[2] - M_PI;
+  std::array<double, 42> jacobian_array = lazyModel()->zeroJacobian(
+    franka::Frame::kEndEffector, q_array, lazy_F_T_EE(), lazy_EE_T_K());
+  return Eigen::Map<const Jacobian>(jacobian_array.data());
+}
 
-  if (angles_norm[1] > M_PI) {
-    angles_norm[1] -= 2 * M_PI;
-  }
-  if (angles_norm[2] < -M_PI) {
-    angles_norm[2] += 2 * M_PI;
-  }
+Vector7d Robot::inverseKinematics(const Affine& target, const Vector7d& q0) {
+    Vector7d q = q0;
+    Matrix7d I = Matrix7d::Identity();
+    const double k0 = 0.01;
+    const int max_iterations = 50;
+    const double tol = 1e-4;
+    const double min_step = 1e-6;
+    const double pinv_reg = 0.1;
 
-  if (angles.norm() < angles_norm.norm()) {
-    angles_norm = angles;
-  }
+    for (int i = 0; i < max_iterations; ++i) {
+        Affine x = this->forwardKinematics(q);
+        Eigen::Matrix<double, 6, 1> e = computeError(x, target);
+
+        double error_norm = e.norm();
+        if (error_norm < tol)
+            break;
 
-  Eigen::Matrix<double, 6, 1> x_target;
-  x_target << target.translation(), angles;
+        Jacobian j = this->jacobian(q);
+        Eigen::Matrix<double, 7, 6> j_inv = j.transpose() * (j * j.transpose() + pinv_reg * Matrix6d::Identity()).inverse();
+
+        Matrix7d N = I - j_inv * j;
```

Comment:
You may add comment specifying the optimization method used here

Reply:
This is a quite standard IK method, though the null space handling may require an additional comment (added)